Update GRMHD analytic data for use in `WrappedGr`

moxcodes opened the pull request:
## Proposed changes

Adds the necessary type aliases and instantiations so that any GRMHD analytic data or Relativistic Euler analytic solution can be used in `WrappedGr` to provide GH+MHD data.

### Upgrade instructions

<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Dependencies

- [x] #3177 (just because of merge conflicts in `WrappedGr`)

moxcodes committed:
Tweak GrMhd Analytic data for WrappedGr compatibility

moxcodes committed:
Tweak RelativisticEuler solutions for WrappedGr compatibility

moxcodes commented:
Updated for the changes in #3177

wthrowe commented on src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/WrappedGr.hpp:
```diff
@@ -94,6 +94,28 @@ class WrappedGr : public SolutionType {
+      tmpl::list<Tags...> /*meta*/) const noexcept {
+    // Get the underlying solution's variables using the solution's tags list,
+    // store in IntermediateVariables
+    const IntermediateVars& intermediate_vars = SolutionType::variables(
```
I assume `SolutionType::variables` will never return a reference, so remove `&` from the variable declaration.  (Again in the next function.)

wthrowe reviewed (changes requested):
The release builds are failing for some reason.

moxcodes commented:
rebased on develop and posted fixups for the requested change and several updates needed to handle recent changes to the analytic solutions and fix mistakes that I found.
Cheers!

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/RelativisticEuler/TovStar.cpp:
```diff
@@ -404,15 +412,39 @@ bool operator!=(const TovStar<RadialSolution>& lhs,
   template bool operator!=(const TovStar<STYPE(data)>& lhs,           \
                            const TovStar<STYPE(data)>& rhs) noexcept;
 
+
```
drop extra line?

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/RelativisticEuler/TovStar.cpp:
```diff
@@ -284,6 +284,14 @@ TovStar<RadialSolution>::variables(
 
+template <typename RadialSolution>
+template <typename DataType, typename Tag>
+tuples::TaggedTuple<::Tags::dt<Tag>> TovStar<RadialSolution>::variables(
```
for my understanding, the `dt` is needed for constructing `Pi`, right?

moxcodes replied:
Yeah, that's right; internally, the `WrappedGr` grabs `dt_shift`, `dt_lapse`, and `dt_spatial_metric` and assembles `Pi`.

nilsdeppe reviewed (commented):
I'm happy for you to squash, including any new changes

wthrowe commented:
I am happy with the fixups so far.

moxcodes commented:
fixups squashed in. This will still need a rebase once the dependency is merged in.
Cheers!

nilsdeppe reviewed (dismissed):
LGTM

nilsdeppe commented:
@moxcodes this needs a rebase now :)

moxcodes commented:
Rebased on develop.
Cheers!

wthrowe commented:
clang-tidy appears to think your code fails to compile, although clang itself disagrees.  I don't know why that would be, but can you check if there is an easy workaround since it strikes me as something that might cause trouble down the road?

moxcodes commented:
Squashed in a fix that at least solved the clang-tidy problem for me locally. There was an inheritance ambiguity that wasn't really ambiguous because one path was `private` and the other `public`, so the actual compilers didn't mind.
I solved it by just not using the base class factoring and added an explanatory comment for the case that clang-tidy didn't like.
Cheers!
